runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java

```diff
@@ -150,41 +154,51 @@ private static boolean canFuseParDo(
+          .values()
+          .contains(candidate.getId())) {
+        // Allow fusion across timer PCollections because they are a self loop.
+        return true;
+      } else if (payload.getStateSpecsCount() > 0 || payload.getTimerSpecsCount() > 0) {
+        // Inputs to a ParDo that uses State or Timers must be key-partitioned, and elements for
+        // a key must execute serially. To avoid checking if the rest of the stage is
+        // key-partitioned and preserves keys, these ParDos do not fuse into an existing stage.
+        return false;
+      } else if (!pipeline.getSideInputs(parDo).isEmpty()) {
+        // At execution time, a Runner is required to only provide inputs to a PTransform that, at
+        // the time the PTransform processes them, the associated window is ready in all side inputs
+        // that the PTransform consumes. For an arbitrary stage, it is significantly complex for the
+        // runner to determine this for each input. As a result, we break fusion to simplify this
+        // inspection. In general, a ParDo which consumes side inputs cannot be fused into an
+        // executable stage alongside any transforms which are upstream of any of its side inputs.
+        return false;
       }
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(e);
     }
     return true;
   }
 
   private static boolean parDoCompatibility(
       PTransformNode parDo, PTransformNode other, QueryablePipeline pipeline) {
-    // This is a convenience rather than a strict requirement. In general, a ParDo that consumes
-    // side inputs can be fused with other transforms in the same environment which are not
-    // upstream of any of the side inputs.
-    return pipeline.getSideInputs(parDo).isEmpty()
-        // Since we lack the ability to mark upstream transforms as key preserving, we
-        // purposefully break fusion here to provide runners the opportunity to insert a
-        // grouping operation
-        && pipeline.getUserStates(parDo).isEmpty()
-        && compatibleEnvironments(parDo, other, pipeline);
+    // Implicitly true if we are attempting to fuse against oneself. This is for timer PCollection which create a loop.
+    return parDo.equals(other)
+        // This is a convenience rather than a strict requirement. In general, a ParDo that consumes
+        // side inputs can be fused with other transforms in the same environment which are not
+        // upstream of any of the side inputs.
+        || (pipeline.getSideInputs(parDo).isEmpty()
+            // Since we lack the ability to mark upstream transforms as key preserving, we
+            // purposefully break fusion here to provide runners the opportunity to insert a
+            // grouping operation
+            && pipeline.getUserStates(parDo).isEmpty()
+            // Since we lack the ability to mark upstream transforms as key preserving, we
```

Comment:
I wanted to make sure that user state and timers had fusion breaks for the same reason. I clarified the comment but still kept it "duplicated" to reiterate this point since it is important.